### PR TITLE
Initial implementation of future<void>::then()

### PR DESCRIPTION
**Pull request by coryan**
This is part of the work for #1345. It implements future<void>::then(),
because future<T> is not implemented yet, it only works for functors
that return `void`.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1391)

### CODE REVIEW

File: google/cloud/internal/future_then_meta.h [hunk 1]
```diff
@@ -97,6 +97,47 @@ struct continuation_helper {
...
 
+/**
+ * A metafunction to implement `future<T>::%then(Functor&&)`.
+ *
+ * This metafunction implements a number of useful results given a functor type
+ * @p Functor, and the value type @p T of a `future<T>`.
+ *
+ * * First it determines if `Functor` meets the requirements, i.e., that it can
+ *   be invoked with an object of type `future<T>` as its only argument.
+ * * Then in computes the type of the expression `functor(fut)`, where `functor`
```
Comment (deepankarsharma) [inline google/cloud/internal/future_then_meta.h]: s/in/it

File: google/cloud/internal/future_void.h [hunk 2]
```diff
@@ -63,10 +64,90 @@ class future<void> final : private internal::future_base<void> {
...
+    // create a non-protected one:
+    using local_state_type = shared_state_type;
+    // Some type aliases to make the rest of the code more readable.
+    using functor_result_t =
+        typename internal::then_helper<F, void>::functor_result_t;
+    using future_t = typename internal::then_helper<F, void>::future_t;
+
+    // The `shared_state_type` (aka `future_shared_state<void>`) is be written
+    // without any reference to the `future<void>` class, otherwise there would
+    // be cycling dependencies between the two classes. We must adapt the
```
Comment (deepankarsharma) [inline google/cloud/internal/future_void.h]: s/cycling/cyclic

File: google/cloud/internal/future_then_meta.h [hunk 3]
```diff
@@ -97,6 +97,47 @@ struct continuation_helper {
...
+ *   is a `future<U>`.
+ * * It computes the type of future returned from `future<T>::%then(Functor&&)`.
+ *
+ * @tparam Functor the functor to call. It must accept a `future<T>` as its
+ *     single input parameter.
+ * @tparam T the type contained in the input future.
+ */
+template <typename Functor, typename T,
+          typename std::enable_if<is_invocable<Functor, future<T>>::value,
+                                  int>::type = 0>
```
Comment (deepankarsharma) [inline google/cloud/internal/future_then_meta.h]: Do we need the int here?

File: google/cloud/internal/future_void.h [hunk 4]
```diff
@@ -63,10 +64,90 @@ class future<void> final : private internal::future_base<void> {
...
  private:
   friend class promise<void>;
   explicit future(std::shared_ptr<shared_state_type> state)
       : future_base<void>(std::move(state)) {}
+
+  /// Implement `then()` if the result does not require unwrapping.
+  template <typename F>
+  typename internal::then_helper<F, void>::future_t then_impl(F&& functor,
+                                                              std::false_type) {
+    // g++-4.9 gets confused about the use of a protected type alias here, so
```
Comment (deepankarsharma) [inline google/cloud/internal/future_void.h]: What is the minimum supported g++ version?

Comment (coryan) [reply google/cloud/internal/future_void.h]: That is poorly defined right now. The CI builds have a gcc-4.9 build enabled.  We decided to support the distributions that GCE supports, which right now include Ubuntu 14.04 and RHEL 7.  Their default compilers are gcc-4.8 based, which we could not get to work, so we sort of settled for 4.9.  We may need to  reconsider that decision (#798).

File: google/cloud/internal/future_then_meta.h (hunk 1 above)
Comment (coryan) [reply google/cloud/internal/future_then_meta.h]: Created #1399 to fix these.

File: google/cloud/internal/future_void.h (hunk 2 above)
Comment (coryan) [reply google/cloud/internal/future_void.h]: ditto.

File: google/cloud/internal/future_then_meta.h (hunk 3 above)
Comment (coryan) [reply google/cloud/internal/future_then_meta.h]: I guess we could say:

`template<..., typename X = typename std::enable_if<predicate>::type>`

but that seems more verbose than:

`template<..., typename std::enable_if<predicate, int>::type = 0>`

Any reason why you think one is better than the other?  I have no strong opinions (other than I want to type less), but would love the learn about a better way to use `std::enable_if<>`.

Comment (deepankarsharma) [reply google/cloud/internal/future_then_meta.h]: I was thinking of something like https://godbolt.org/z/HzNegx

**Pull request merged**